### PR TITLE
Scale Testing

### PR DESCRIPTION
**Pull request by kohsheen1234**
RFC for Scale Test (Step 1)

Rendered copy of RFC : https://github.com/kohsheen1234/gsoc/blob/master/rfc/2020/scale-testing/kohsheen1234/rfc.md

### CODE REVIEW

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+- Start Date: 29-02-2020
+- Target Date: 
+- RFC PR:
+- Linkerd Issue: [linkerd/linkerd2#3895](https://github.com/linkerd/linkerd2/issues/3895)
+- Reviewers: 
+
+[summary]: #summary
+
+# Summary
+This conrtibution aims to completely automate the scale tests, so that scale-testing can be performed very easily. Automating scale tests is done to ensure the performance of Linkerd is maintained.
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: s/automate the scale tests/automate the scale test framework
s/scale-testing/scale testing

> Automating scale tests is done to ensure the performance of Linkerd is maintained.

The reason why we want to automate the scale test framework is to ensure the tests are repeatable by the community. The test suite doesn't _maintain_ the performance of Linkerd. Its serves to provide visibility into the proxy's performance and reveal any performance degration.

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+# Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+- Identify performance regression/potential errors with minimal manual intervention.
+- Latency in the network communication can be a massive issue. When new features added to Linkerd, it might detoirate the performance of Linkerd. So, it is important to understand if any new feature which has been added recently has a negative impact on Linkerd's performance. Manually checking every-time if there has been a regression in the performance is not a convinient solution. This process can be automated, so that it is very simple to perform scale testing.
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: s/detoirate/deteriorate
s/every-time/everyt time
s/convinent/convenient

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+- Once this scale testing has been automated, Linkerd will be able to :
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: > Linkerd will be able to

Linkerd *will not* be adding a sample workload or report on any resource usage. Your new test framework will.

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  - Automatically add a sample workload to the cluster.
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md:20]: In phase 2, consider whether your test framework will automate the installation of Linkerd and other infrastructure resources like Kubernetes cluster, tainted nodes to run Prometheus etc.

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+# Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- Scale and Performance testing is included in other service mesh.
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: s/Performance/performance

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- Scale and Performance testing is included in other service mesh. 
+  - For instance, let's consider Istio https://istio.io/docs/ops/deployment/performance-and-scalability/. 
+  - Github link - https://github.com/istio/tools/tree/release-1.4/perf
+- Current scale-test of linkerd - https://github.com/linkerd/linkerd2/blob/master/bin/test-scale
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md:41]: s/linkerd/Linkerd

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+- Identify performance regression/potential errors with minimal manual intervention.
+- Latency in the network communication can be a massive issue. When new features added to Linkerd, it might detoirate the performance of Linkerd. So, it is important to understand if any new feature which has been added recently has a negative impact on Linkerd's performance. Manually checking every-time if there has been a regression in the performance is not a convinient solution. This process can be automated, so that it is very simple to perform scale testing.
+- Once this scale testing has been automated, Linkerd will be able to :
+  - Automatically add a sample workload to the cluster.
+  - Record cluster,control pane and data plane metrics during the test.
+    - Success rate
+    - Latency and throughput
+    - Dashboard load times
+    - Linkerd stat responsiveness
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: Add backticks to _linkerd stat_ since it's a CLI command.

**Review by ihcsim — COMMENTED** [review]
@kohsheen1234 Thanks for this RFC. See comments below. Once addressed, feel free to proceed to phase 2.

**Comment by kohsheen1234** [comment]
@ihcsim Thank you for reviewing, I have made the changes as suggested and will now move on to phase 2. 🙂 :+1:

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+- Start Date: 29-02-2020
+- Target Date: 
+- RFC PR:
+- Linkerd Issue: [linkerd/linkerd2#3895](https://github.com/linkerd/linkerd2/issues/3895)
+- Reviewers: 
+
+[summary]: #summary
+
+# Summary
+This conrtibution aims to completely automate the scale test framework, to ensure the scale tests are repeatable by the community. Automating scale tests serve to provide visibility into the proxy's performance and reveal any performance degration.
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: s/conrtibution/contribution

File: rfc/2020/scale-testing/kohsheen1234/rfc.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+- Identify performance regression/potential errors with minimal manual intervention.
+- Latency in the network communication can be a massive issue. When new features added to Linkerd, it might deteriorate the performance of Linkerd. So, it is important to understand if any new feature which has been added recently has a negative impact on Linkerd's performance. Manually checking every time if there has been a regression in the performance is not a convenient solution. This process can be automated, so that it is very simple to perform scale testing.
+- Once this scale testing has been automated,the scale test framework will be able to :
+  - Automatically add a sample workload to the cluster.
+  - Record cluster,control pane and data plane metrics during the test.
+    - Success rate
+    - Latency and throughput
+    - Dashboard load times
+    - Add backticks to linkerd stat
```
Comment (ihcsim) [inline rfc/2020/scale-testing/kohsheen1234/rfc.md]: What I meant is this line should look like (notice the formatting of 'linkerd stat'):

- `linkerd stat` responsiveness

Comment (kohsheen1234) [reply rfc/2020/scale-testing/kohsheen1234/rfc.md]: Sorry, very silly of me 😝

**Pull request closed without merging**